File: matlab/src/solvers/MVA/solver_mva_oi_analyzer.m
function [QN,UN,RN,TN,CN,XN,lG,runtime,iter,actualmethod] = solver_mva_oi_analyzer(sn, options)
% SOLVER_MVA_OI_ANALYZER Exact mean-value MVA for order-independent networks
%
% An order-independent (OI) station is a class-dependent load-dependent server
% whose total service rate mu(n) is a permutation-invariant function of the
% per-class count vector n. A closed network of infinite-server (delay) and
% load-independent (single-server, product-form) stations plus ANY number of OI
% stations is product-form. This analyzer aggregates the delay stations into a
% single think-time vector Z, collects the load-independent (LI) queue demands
% and the OI-station rate handles, and calls PFQN_MVAOI, the mean-value
% Conditional-MVA (CMVA) that carries one rate-shift vector per OI station and
% returns exact per-class throughput and queue-lengths WITHOUT any normalizing
% constant or joint marginal. The marginal-distribution counterpart is
% PFQN_MVAOI_MARG.
%
% Reference:
%   Reiser, Lavenberg (1980). Mean-Value Analysis of Closed Multichain Queuing
%   Networks. JACM 27(2). Load-dependent extension: Bruell, Balbo, Afshari
%   (1984). OI stations / CMVA: Casale (2009); Casale, Comte, Dorsman (2026).
%
% Copyright (c) 2012-2026, Imperial College London
% All rights reserved.

tstart = tic;

oi_list = find_oi_stations(sn);
if isempty(oi_list)
    error('solver_mva_oi_analyzer:NoOIStation', ...
        'OI solver requires at least one order-independent station');
end

% ---- reject class switching (OI rank rates are per raw class) --------------
% The recursion is driven by the per-class population vector sn.njobs, which
% class switching makes meaningless: a class that only ever appears mid-chain
% carries njobs = 0, so the OI station would be analyzed as if empty. Refuse it
% the way SOLVER_NC_OI_ANALYZER does rather than return that silently.
for c = 1:sn.nchains
    if numel(sn.inchain{c}) > 1
        line_error(mfilename, 'solver_mva_oi requires one class per chain (no class switching).');
    end
end

M = sn.nstations;
R = sn.nclasses;
N = round(sn.njobs(:)');

isOI = false(1, M);
isOI(oi_list) = true;
isDelay = (sn.sched(:)' == SchedStrategy.INF);

% Per-class service demand D_ir = V_ir/rate_ir at non-OI stations.
V = zeros(M, R);
for c = 1:numel(sn.visits)
    V = V + sn.visits{c};
end
D = zeros(M, R);
for i = 1:M
    for r = 1:R
        if isfinite(sn.rates(i,r)) && sn.rates(i,r) > 0
            D(i,r) = V(i,r) / sn.rates(i,r);
        end
    end
end

% Aggregate the delay stations into Z; collect the LI queue demands and OI-station
% rate handles. A multiserver BCMP queue (c>1) cannot take the single-server LI
% path, so it is promoted to an OI station via the c-server BCMP weight; see
% _kb/06-solver-catalog.md (NC section, OI analyzer) for the W(n)/mu(n) identity.
Z = zeros(1, R);
li_list = [];
ms_list = [];
for i = 1:M
    if isOI(i)
        continue
    elseif isDelay(i)
        Z = Z + D(i,:);
    elseif isfinite(sn.nservers(i)) && sn.nservers(i) > 1
        ms_list(end+1) = i; %#ok<AGROW>
    else
        li_list(end+1) = i; %#ok<AGROW>
    end
end
Dli = D(li_list, :);

muCell = cell(1, numel(oi_list) + numel(ms_list));
% Per-muCell-station visit vectors. Genuine OI stations carry their class visit
% ratios V(oi,:) (the rate handle has no visits); ms-promoted stations pass
% ones, as their visits are already folded into D by ms_oi_rate.
oivis = cell(1, numel(oi_list) + numel(ms_list));
for o = 1:numel(oi_list)
    node_oi = sn.stationToNode(oi_list(o));
    svcRateFun = sn.nodeparam{node_oi}.svcRateFun;
    muCell{o} = @(n) oi_rate(svcRateFun, n, R);
    oivis{o} = V(oi_list(o), :);
end
for j = 1:numel(ms_list)
    muCell{numel(oi_list) + j} = ms_oi_rate(D(ms_list(j), :), sn.nservers(ms_list(j)));
    oivis{numel(oi_list) + j} = ones(1, R);
end

[X, Qoi, Qli, ~, Soi] = pfqn_mvaoi(Z, N, muCell, Dli, oivis, options);

% Assemble per-station mean queue-lengths.
QN = zeros(M, R);
for o = 1:numel(oi_list)
    QN(oi_list(o), :) = Qoi(o, :);
end
for j = 1:numel(ms_list)
    QN(ms_list(j), :) = Qoi(numel(oi_list) + j, :);
end
for j = 1:numel(li_list)
    QN(li_list(j), :) = Qli(j, :);
end
for i = 1:M
    if isDelay(i)
        QN(i, :) = X .* D(i, :);     % IS station: exact product-form share
    end
end
XN = X;

% Row of Soi/Qoi holding each OI station (muCell order: oi_list, then ms_list).
oiRow = zeros(1, M);
for o = 1:numel(oi_list)
    oiRow(oi_list(o)) = o;
end

% Assemble outputs at the full population.
TN = zeros(M, R);
RN = zeros(M, R);
UN = zeros(M, R);
CN = zeros(M, R);
for i = 1:M
    for r = 1:R
        TN(i,r) = XN(r) * V(i,r);
        if XN(r) > 0
            RN(i,r) = QN(i,r) / XN(r);
        end
        if isOI(i)
            % In-service utilization U_r = E[sir_r]/nservers; see
            % _kb/06-solver-catalog.md (NC section, OI analyzer). Soi holds E[sir_r].
            sv = sn.nservers(i);
            if ~isfinite(sv) || sv <= 0
                sv = 1;
            end
            UN(i,r) = Soi(oiRow(i), r) / sv;
        elseif isDelay(i)
            UN(i,r) = QN(i,r);
        else
            sv = sn.nservers(i);
            if ~isfinite(sv) || sv <= 0
                sv = 1;
            end
            UN(i,r) = XN(r) * D(i,r) / sv;
        end
        CN(i,r) = RN(i,r);
    end
end

lG = 0;
runtime = toc(tstart);
iter = sum(N);
actualmethod = 'oi';
end

% =========================================================================
% Helper functions
% =========================================================================

function oi_list = find_oi_stations(sn)
% Station indices of all OI stations: PAS/OI scheduling with an all-zero swap
% graph and a service-rate function (mirrors nc_is_oi_model detection).
oi_list = [];
for ist = 1:sn.nstations
    if sn.sched(ist) ~= SchedStrategy.PAS && sn.sched(ist) ~= SchedStrategy.OI
        continue
    end
    ind = sn.stationToNode(ist);
    if ind < 1 || ind > numel(sn.nodeparam) || ~isstruct(sn.nodeparam{ind})
        continue
    end
    np = sn.nodeparam{ind};
    if ~isfield(np, 'swapGraph') || ~isfield(np, 'svcRateFun') || isempty(np.svcRateFun)
        continue
    end
    sg = np.swapGraph;
    if isempty(sg) || any(sg(:) ~= 0)
        continue
    end
    oi_list(end+1) = ist; %#ok<AGROW>
end
end

function h = ms_oi_rate(Dq, c)
% OI rate function reproducing the c-server BCMP station with per-class demands
% Dq: mu(n) = (min(|n|,c)/|n|) * sum_{r: n_r>0} n_r/Dq_r. For c = 1 this is the
% familiar total completion rate of a multiclass single-server queue, and for a
% single class it reduces to min(n,c)/Dq (M/M/c).
Dq = Dq(:)';
if ~isfinite(c) || c <= 0
    c = 1;
end
h = @(n) ms_oi_rate_eval(n, Dq, c);
end

function rate = ms_oi_rate_eval(n, Dq, c)
tot = sum(n);
if tot == 0
    rate = 0;
    return
end
acc = 0;
for r = 1:numel(n)
    if n(r) > 0 && Dq(r) > 0
        acc = acc + n(r) / Dq(r);
    end
end
rate = (min(tot, c) / tot) * acc;
end

function rate = oi_rate(svcRateFun, n, R)
% OI total service rate at count vector n via the 1-based canonical microstate.
if sum(n) == 0
    rate = 0;
    return
end
micro = repelem(1:R, n);
rate = svcRateFun(micro);
end

File: jar/src/main/java/jline/solvers/mva/SolverMVAOIAnalyzer.java
package jline.solvers.mva;

import jline.api.pfqn.mva.Pfqn_mvaoi;
import jline.lang.NetworkStruct;
import jline.lang.constant.SchedStrategy;
import jline.lang.nodeparam.QueueNodeParam;
import jline.lang.nodes.Station;
import jline.solvers.SolverOptions;
import jline.util.SerializableFunction;
import jline.util.matrix.Matrix;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.function.ToDoubleFunction;

/**
 * Exact mean-value MVA for order-independent (OI) queueing networks.
 *
 * An OI station is a class-dependent load-dependent server whose total service
 * rate mu(n) is a permutation-invariant function of the per-class count vector n.
 * A closed network of infinite-server (delay) and load-independent (single-server,
 * product-form) stations plus ANY number of OI stations is product-form. This
 * analyzer aggregates the delay stations into a single think-time vector Z,
 * collects the load-independent (LI) queue demands and the OI-station rate
 * handles, and calls {@link Pfqn_mvaoi}, the mean-value Conditional-MVA (CMVA)
 * that carries one rate-shift vector per OI station and returns exact per-class
 * throughput and queue-lengths WITHOUT any normalizing constant or joint
 * marginal. The marginal-distribution counterpart is
 * {@link jline.api.pfqn.mva.Pfqn_mvaoi_marg}.
 *
 * Reference: Reiser, Lavenberg (1980), JACM 27(2); load-dependent extension Bruell,
 * Balbo, Afshari (1984); OI stations / CMVA Casale (2009); Casale, Comte, Dorsman (2026).
 */
public class SolverMVAOIAnalyzer {

    private final NetworkStruct sn;
    private final SolverOptions options;
    private final int R;
    private final int M;
    private final int[] njobs;
    private final double[][] demand;   // D_ir
    private final double[][] visits;   // V_ir
    private final boolean[] isDelay;
    private final boolean[] isOI;
    private final List<Integer> oiList;

    public SolverMVAOIAnalyzer(NetworkStruct sn, SolverOptions options) {
        this.sn = sn;
        this.options = options;
        this.R = sn.nclasses;
        this.M = sn.nstations;
        this.njobs = new int[R];
        for (int r = 0; r < R; r++) {
            this.njobs[r] = (int) Math.round(sn.njobs.get(0, r));
        }
        this.oiList = findOIStations();
        if (this.oiList.isEmpty()) {
            throw new RuntimeException("OI solver requires at least one order-independent station");
        }
        // ---- reject class switching (OI rank rates are per raw class) ------
        // The recursion is driven by the per-class population vector sn.njobs,
        // which class switching makes meaningless: a class that only ever
        // appears mid-chain carries njobs = 0, so the OI station would be
        // analyzed as if empty. Refuse it the way Solver_nc_oi does rather than
        // return that silently.
        for (int c = 0; c < sn.nchains; c++) {
            if (sn.inchain.get(c).getNumElements() > 1) {
                throw new RuntimeException("solver_mva_oi requires one class per chain (no class switching).");
            }
        }
        this.visits = new double[M][R];
        for (Map.Entry<Integer, Matrix> e : sn.visits.entrySet()) {
            Matrix vc = e.getValue();
            for (int i = 0; i < M; i++) {
                for (int r = 0; r < R; r++) {
                    this.visits[i][r] += vc.get(i, r);
                }
            }
        }
        this.demand = new double[M][R];
        for (int i = 0; i < M; i++) {
            for (int r = 0; r < R; r++) {
                double rate = sn.rates.get(i, r);
                if (!Double.isInfinite(rate) && !Double.isNaN(rate) && rate > 0) {
                    this.demand[i][r] = this.visits[i][r] / rate;
                }
            }
        }
        this.isDelay = new boolean[M];
        this.isOI = new boolean[M];
        for (int i = 0; i < M; i++) {
            this.isDelay[i] = sn.sched.get(sn.stations.get(i)) == SchedStrategy.INF;
        }
        for (int o : this.oiList) this.isOI[o] = true;
    }

    private List<Integer> findOIStations() {
        List<Integer> out = new ArrayList<Integer>();
        for (int i = 0; i < M; i++) {
            Station st = sn.stations.get(i);
            SchedStrategy s = sn.sched.get(st);
            if (s != SchedStrategy.PAS && s != SchedStrategy.OI) {
                continue;
            }
            Object p = sn.nodeparam.get(st);
            if (!(p instanceof QueueNodeParam)) {
                continue;
            }
            QueueNodeParam qp = (QueueNodeParam) p;
            if (qp.svcRateFun == null || qp.swapGraph == null) {
                continue;
            }
            if (qp.swapGraph.isEmpty() || qp.swapGraph.elementMaxAbs() == 0) {
                out.add(i);
            }
        }
        return out;
    }

    /** OI total service rate at count vector n via the (0-based) canonical microstate. */
    private static double muM(SerializableFunction<Matrix, Double> svcRateFun, int[] n, int R) {
        int tot = 0;
        for (int v : n) tot += v;
        if (tot == 0) {
            return 0.0;
        }
        Matrix micro = new Matrix(1, tot);
        int pos = 0;
        for (int r = 0; r < R; r++) {
            for (int c = 0; c < n[r]; c++) {
                micro.set(0, pos++, r);
            }
        }
        return svcRateFun.apply(micro);
    }

    /**
     * OI rate function reproducing the c-server BCMP station with per-class demands
     * Dq: mu(n) = (min(|n|,c)/|n|) * sum_{r: n_r&gt;0} n_r/Dq_r. For c = 1 this is the
     * familiar total completion rate of a multiclass single-server queue, and for a
     * single class it reduces to min(n,c)/Dq (M/M/c).
     */
    private static double msOIRate(int[] n, double[] Dq, double c) {
        int tot = 0;
        for (int v : n) tot += v;
        if (tot == 0) {
            return 0.0;
        }
        double acc = 0.0;
        for (int r = 0; r < n.length; r++) {
            if (n[r] > 0 && Dq[r] > 0) {
                acc += n[r] / Dq[r];
            }
        }
        return (Math.min(tot, c) / tot) * acc;
    }

    public AnalysisResults analyze() {
        long tstart = System.currentTimeMillis();

        // see _kb/06-solver-catalog.md for rationale
        double[] Z = new double[R];
        List<Integer> liList = new ArrayList<Integer>();
        List<Integer> msList = new ArrayList<Integer>();
        for (int i = 0; i < M; i++) {
            if (isOI[i]) {
                continue;
            } else if (isDelay[i]) {
                for (int r = 0; r < R; r++) Z[r] += demand[i][r];
            } else {
                double sv = sn.nservers.get(i, 0);
                if (!Double.isInfinite(sv) && !Double.isNaN(sv) && sv > 1) {
                    msList.add(i);
                } else {
                    liList.add(i);
                }
            }
        }
        double[][] Dli = new double[liList.size()][R];
        for (int j = 0; j < liList.size(); j++) {
            Dli[j] = demand[liList.get(j)].clone();
        }

        List<ToDoubleFunction<int[]>> muCell = new ArrayList<ToDoubleFunction<int[]>>();
        for (int o : oiList) {
            Station oiStation = sn.stations.get(o);
            QueueNodeParam qp = (QueueNodeParam) sn.nodeparam.get(oiStation);
            final SerializableFunction<Matrix, Double> f = qp.svcRateFun;
            muCell.add(new ToDoubleFunction<int[]>() {
                public double applyAsDouble(int[] n) {
                    return muM(f, n, R);
                }
            });
        }
        for (int j = 0; j < msList.size(); j++) {
            final double[] Dq = demand[msList.get(j)].clone();
            double svTmp = sn.nservers.get(msList.get(j), 0);
            if (Double.isInfinite(svTmp) || Double.isNaN(svTmp) || svTmp <= 0) {
                svTmp = 1;
            }
            final double cSv = svTmp;
            muCell.add(new ToDoubleFunction<int[]>() {
                public double applyAsDouble(int[] n) {
                    return msOIRate(n, Dq, cSv);
                }
            });
        }

        // Per-muCell-station visit vectors: genuine OI stations carry their
        // class visits (rate handle has none); ms-promoted stations pass unit
        // visits (already folded into demand by msOIRate).
        double[][] oivis = new double[oiList.size() + msList.size()][R];
        for (int o = 0; o < oiList.size(); o++) {
            for (int r = 0; r < R; r++) oivis[o][r] = visits[oiList.get(o)][r];
        }
        for (int j = 0; j < msList.size(); j++) {
            for (int r = 0; r < R; r++) oivis[oiList.size() + j][r] = 1.0;
        }
        Pfqn_mvaoi.Result res = Pfqn_mvaoi.pfqn_mvaoi(Z, njobs, muCell, Dli, oivis);

        double[] XN = res.X;

        // Row of res.Soi/res.Qoi holding each OI station (muCell order: oiList, msList).
        int[] oiRow = new int[M];
        for (int o = 0; o < oiList.size(); o++) oiRow[oiList.get(o)] = o;
        double[][] QN = new double[M][R];
        for (int o = 0; o < oiList.size(); o++) {
            QN[oiList.get(o)] = res.Qoi[o].clone();
        }
        for (int j = 0; j < msList.size(); j++) {
            QN[msList.get(j)] = res.Qoi[oiList.size() + j].clone();
        }
        for (int j = 0; j < liList.size(); j++) {
            QN[liList.get(j)] = res.Qli[j].clone();
        }
        for (int i = 0; i < M; i++) {
            if (isDelay[i]) {
                for (int r = 0; r < R; r++) QN[i][r] = XN[r] * demand[i][r];
            }
        }

        double[][] TN = new double[M][R];
        double[][] RN = new double[M][R];
        double[][] UN = new double[M][R];
        double[][] CN = new double[M][R];
        for (int i = 0; i < M; i++) {
            for (int r = 0; r < R; r++) {
                TN[i][r] = XN[r] * visits[i][r];
                if (XN[r] > 0) {
                    RN[i][r] = QN[i][r] / XN[r];
                }
                if (isOI[i]) {
                    // see _kb/06-solver-catalog.md for rationale
                    double sv = sn.nservers.get(i, 0);
                    if (Double.isInfinite(sv) || sv <= 0) sv = 1;
                    UN[i][r] = res.getSoi()[oiRow[i]][r] / sv;
                } else if (isDelay[i]) {
                    UN[i][r] = QN[i][r];
                } else {
                    double sv = sn.nservers.get(i, 0);
                    if (Double.isInfinite(sv) || sv <= 0) sv = 1;
                    UN[i][r] = XN[r] * demand[i][r] / sv;
                }
                CN[i][r] = RN[i][r];
            }
        }

        int Ntot = 0;
        for (int v : njobs) Ntot += v;
        long runtime = System.currentTimeMillis() - tstart;
        return new AnalysisResults(QN, UN, RN, TN, CN, XN, 0.0, runtime, Ntot, "oi");
    }

    public static class AnalysisResults {
        public double[][] QN, UN, RN, TN, CN;
        public double[] XN;
        public double lG;
        public long runtime;
        public int iter;
        public String method;

        public AnalysisResults(double[][] QN, double[][] UN, double[][] RN,
                double[][] TN, double[][] CN, double[] XN,
                double lG, long runtime, int iter, String method) {
            this.QN = QN;
            this.UN = UN;
            this.RN = RN;
            this.TN = TN;
            this.CN = CN;
            this.XN = XN;
            this.lG = lG;
            this.runtime = runtime;
            this.iter = iter;
            this.method = method;
        }
    }
}
